### PR TITLE
Issue #114: Get all bookings of an amenity

### LINKED ISSUE #114: Get all bookings for an amenity with an optional time range
## ⚠️ Is your feature request related to a problem? Please describe

* Get all bookings for an amenity with an optional time range

## 💡 Describe the solution you'd like

* `GET /amenities/{amenityId}/bookings?start='01-01-2020'&end='01-02-2020'`
* The API should support pagination
* Code pointers for pagination - `CommunityService.java::listAll`
* Update postman collection

## Task Relationship

Subtask of: jmprathab/MyHome#94

### PR DESCRIPTION
**Pull request by nlaert**
## 🚀 Description
 
Add the possibility to get all bookings of one amenity, with optional time range

## 🧪 How Has This Been Tested?
Unit Test for controller + postman call

## 📦 Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] My code follows the code style of this project(Do your best to follow code styles. If none apply just skip this).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/com/myhome/controllers/response/GetAmenityBookingsResponse.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
+public class GetAmenityBookingsResponse {
+    String amenityId;
```
Comment (mslowiak) [inline src/main/java/com/myhome/controllers/response/GetAmenityBookingsResponse.java]: That fields could be private and class annotated as lombok Data annotation

File: src/test/java/com/myhome/controllers/unit/AmenityControllerTest.java [hunk 2]
```diff
@@ -154,9 +162,46 @@ void shouldNotAddAmenityWhenCommunityNotExists() {
     assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
   }
 
+  @Test
+  void shouldReturnListOfBookingsOfGivenAmenityId() {
```
Comment (mslowiak) [inline src/test/java/com/myhome/controllers/unit/AmenityControllerTest.java]: Can you apply given/when/then pattern to tests?

File: src/main/java/com/myhome/controllers/AmenityController.java [hunk 3]
```diff
@@ -87,4 +97,22 @@ public ResponseEntity deleteAmenity(@PathVariable String amenityId) {
...
+  public ResponseEntity<List<GetAmenityBookingsResponse>> getAmenitiesBookings(
+          @PathVariable String amenityId,
+          @RequestParam(required = false) LocalDateTime start,
+          @RequestParam(required = false) LocalDateTime end,
+          @PageableDefault(size = 200) Pageable pageable) {
+
+    List<AmenityBookingItem> amenityBookingItems =
+            amenitySDJpaService.listAllAmenityBookings(amenityId, start, end, pageable);
+    if (amenityBookingItems.isEmpty()) {
+      return ResponseEntity.notFound().build();
```
Comment (mslowiak) [inline src/main/java/com/myhome/controllers/AmenityController.java]: 404 not found should be returned only when `amenityId` not exists.
In case when there is `amenityId` but the query returns no results (there are no bookings for that amenity) we should return an empty array with 200 code

**Review by mslowiak — CHANGES_REQUESTED** [review]
Please take a look on codestyle that we use in project and reformat created code with it.
Resolve conflicts as well. 

More info about code style is here: https://github.com/jmprathab/MyHome/blob/master/CONTRIBUTING.md#coding-style

File: src/test/java/com/myhome/controllers/unit/AmenityControllerTest.java [hunk 4]
```diff
@@ -154,9 +162,46 @@ void shouldNotAddAmenityWhenCommunityNotExists() {
...
+            amenityController.getAmenitiesBookings(TEST_AMENITY_ID, currentDate, currentDate.plusDays(5), null);
+
+    assertEquals(HttpStatus.OK, amenitiesBookings.getStatusCode());
+    assertNotNull(amenitiesBookings.getBody());
+    assertEquals(1, amenitiesBookings.getBody().size());
+  }
+
+  @Test
+  void shouldReturn404WhenNoBookingsFound() {
+    LocalDateTime currentDate = LocalDateTime.now();
```
Comment (mslowiak) [inline src/test/java/com/myhome/controllers/unit/AmenityControllerTest.java]: Can you add test cases when startDate or endDate is null? What will happen then?

File: src/main/java/com/myhome/controllers/AmenityController.java (hunk 3 above)
Comment (nlaert) [reply src/main/java/com/myhome/controllers/AmenityController.java]: done, just have one doubt. this way it will never return a 404 unless we do another query on the service when the list is empty to check if the amenityId exists. do you think it makes sense?

File: src/test/java/com/myhome/controllers/unit/AmenityControllerTest.java (hunk 2 above)
Comment (nlaert) [reply src/test/java/com/myhome/controllers/unit/AmenityControllerTest.java]: done

File: src/main/java/com/myhome/controllers/response/GetAmenityBookingsResponse.java (hunk 1 above)
Comment (nlaert) [reply src/main/java/com/myhome/controllers/response/GetAmenityBookingsResponse.java]: done

**Review by nlaert — COMMENTED** [review]
Reformated all files with according style. Sorry for not noticing it

File: src/main/java/com/myhome/controllers/AmenityController.java (hunk 3 above)
Comment (mslowiak) [reply src/main/java/com/myhome/controllers/AmenityController.java]: Yes you are right, we will have then additional call, we have two ways:
- create HEAD endpoint which will do that operation (204 if resource exists, 404 if not exists) and we will call HEAD and GET requests on frontend side
- add one more database call inside GET endpoint which will cover that situation (if no records with given id, then return 404)

@jmprathab  What is your suggestion in this topic?

File: service/src/test/java/com/myhome/controllers/AmenityControllerTest.java [hunk 5]
```diff
@@ -213,6 +222,152 @@ void shouldNotUpdateCommunityAmenityIfAmenityNotExists() {
...
+    // when
+    ResponseEntity<List<GetAmenityBookingsResponse>> amenitiesBookings =
+        amenityController.getAmenitiesBookings(TEST_AMENITY_ID,
+            null,
+            futureTime,
+            null);
+
+    // then
+    assertEquals(HttpStatus.OK, amenitiesBookings.getStatusCode());
+    assertNotNull(amenitiesBookings.getBody());
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/controllers/AmenityControllerTest.java]: `amenitiesBookings.getBody()` can be assigned to variable (in test context) as it is used twice:
```
    assertEquals(HttpStatus.OK, amenitiesBookings.getStatusCode());
    // here
    assertNotNull(amenitiesBookings.getBody());
    assertEquals(1, amenitiesBookings.getBody().size());
```

**Review by mslowiak — CHANGES_REQUESTED** [review]

**Comment by mslowiak** [comment]
> Reformated all files with according style. Sorry for not noticing it

Great job, let's wait for @jmprathab to comment on 404 topic :)

File: src/main/java/com/myhome/controllers/AmenityController.java (hunk 3 above)
Comment (jmprathab) [reply src/main/java/com/myhome/controllers/AmenityController.java]: @nlaert @mslowiak My suggestion is to handle it internally as part of a `GET` endpoint.

**Comment by jmprathab** [comment]
@nlaert Please resolve conflicts.

**Comment by nlaert** [comment]
Hello, Added the changes and fixed some conflicts but now there are more conflicts and it's getting late here. 
PR to be continued

File: service/src/test/java/com/myhome/controllers/AmenityControllerTest.java (hunk 5 above)
Comment (nlaert) [reply service/src/test/java/com/myhome/controllers/AmenityControllerTest.java]: done

File: src/main/java/com/myhome/controllers/AmenityController.java (hunk 3 above)
Comment (nlaert) [reply src/main/java/com/myhome/controllers/AmenityController.java]: done, thanks for the input

**Comment by mslowiak** [comment]
Hi,
The commits history looks awful, please fix it.
If you need some help with that just ping me on our discord channel

**Comment by jmprathab** [comment]
@nlaert waiting for your action on fixing the commits as mentioned above by @mslowiak

**Comment by nlaert** [comment]
Hello, sorry I haven't had much time to do this. Tried to squash some commits but it feels so much work for the amount of the code I did. Would it be ok to just delete this PR and create a new one?

**Comment by mslowiak** [comment]
> Hello, sorry I haven't had much time to do this. Tried to squash some commits but it feels so much work for the amount of the code I did. Would it be ok to just delete this PR and create a new one?

If it is easier for you in that way feel free to do it.
Please take a look on https://github.com/jmprathab/MyHome/pull/204#issuecomment-719978271 as well.
Waiting for the new PR @nlaert 👍
